text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,282 @@
+   various `rustc_*` crates there is no intention of keeping around.
+
+ - Cargo doesn't currently have any unstable features, so giving them that requires work and
+   planning.
+
+ - Only some crates in the rust repo (at least `core`, `alloc` and `collections`) can properly be
+   built just based upon their `Cargo.toml`. However it's precisely only these "foundational" crates
+   that will be of interest to freestanding developers. Hosted developers can get pre-built sysroots
+   with `rustup`, and those working on the standard library would use this in conjunction with its
+   build system which can deal with such things.
+
+ - No means of compiling compiler-rt is proposed. But just as freestanding developers need to
+   provide `rlibc` or similar to successfully link, I think that for the time-being they can be left
+   to provide this themselves when linking.
+
+
+# Alternatives
+
+ - It is possible we can just use wildcard dependencies for standard library crates, and don't need
+   to add a notion of a standard library dependency. Crucially, while there are many different
+   versions of the standard library, there is only one that works with any given version of `rustc`.
+
+   Specifically, we can have it so whenever Cargo comes across a wildcard crates.io dependency it
+   can't resolve, assume it is a std lib dependency and either resolve it as such (desugar to git
+   dependencies) or leave it for `rustc` to find in the sysroot according to the build type.
+
+   I made this to be an alternative because it might be a bit too magical. Also, Semver won't be
+   able to understand the release trains---pre-release/beta builds are allowed in its grammar but
+   have different semantics.
+
+ - Simply have `implicit-deps = false` make Cargo pass `--use-sysroot=false` to `rustc`, and don't
+   have any way to explicitly depend on things in the sysroot.
+
+   - This doesn't by-itself make a way for package to depend on only some of the crates behind the
+     facade. That, in turn, means Cargo is little better at cross compiling those than before.
+
+   - While unstable compiler users can just package the standard library and depend on it as a
+     normal crate, it would be weird to have freestanding projects coalesce around some bootleg
+     core on crates.io.
+
+ - Make it so that packages with implicit dependencies only depend on std. This would be more
+   elegant, but breaks packages (including ones using stable Rust) that just depend on crates.io.
+
+ - Make it so all dependencies, even libstd, must be explicit. C.f. Cabal and base. Simple and
+   elegant, but breaks all existing packages.
+
+ - Don't do this, and be stuck with problems detailed in the motivation section.
+
+
+# Unresolved questions
+
+ - There are multiple lists of dependencies for different things (e.g. tests, build-time). How
+   should `implicit-deps = false` affect them?
```

Comment:
Also perhaps I should match `[dependencies]` and `[dev-dependencies]` using those keywords? Is there are third sort (I seem to recall there was)?